Added task for cleaning old tags , and ability to set number of tags to be kept

bogdanRada opened the pull request:
Hello, 
  As per discussion here #9 i added a task for cleaning old tags, and ability to set number of tags to be kept . 
  @technicalpickles  or @jnewland please let me know what you think. I have tested it locally and seems to work fine. 
  Let me know if there is anything that needs to be changed or improved. 
Or if anyone has any suggestion i am opened to discussion. 
Thank you very much.

zocoi commented:
this looks good to me, can we have a ship it?

bogdanRada commented:
well, i guess so. i was trying to get an opinion from other collaborators , just to see if this could be improved in some way. I will wait though until monday because i am currently travelling so i can't do changes from my phone . If everything is ok , will release it then. Sounds good?

technicalpickles commented on lib/capistrano/gitflow/helpers/helper.rb:
```diff
@@ -216,5 +216,23 @@ def gitflow_tag_production
+        if tags_to_delete.any?
+          tags_without_refs = tags_to_delete.map{|tag| tag.gsub('refs/tags/', '') }
+          system "echo #{tags_without_refs.join(' ')} | xargs git tag -d"
+          system "echo #{tags_to_delete.join(' ')} | tr ' ' '\n'  | awk '{print \":\" $0}' | xargs git push origin "
```
It might be simpler to add the `:` to the tag list before making the system call. That way, you don't have to `tr` and `awk` them, and just pass the list (with :) to `git push origin`

technicalpickles commented on lib/capistrano/gitflow/helpers/helper.rb:
```diff
@@ -216,5 +216,23 @@ def gitflow_tag_production
+        tags_to_delete = (tags - tags.last(fetch(:gitflow_keep_tags)))
+        if tags_to_delete.any?
+          tags_without_refs = tags_to_delete.map{|tag| tag.gsub('refs/tags/', '') }
+          system "echo #{tags_without_refs.join(' ')} | xargs git tag -d"
```
I think you can get away with just passing `tags_without_refs.join(' ')` to `git tag -d` rather than using `xargs`

bogdanRada replied:
"git tag -d" accepts only one tag as parameter, that's why i used "xargs" to simplify a bit the code, so that i don't have to use a loop statement,

technicalpickles commented on lib/capistrano/gitflow/helpers/helper.rb:
```diff
@@ -216,5 +216,23 @@ def gitflow_tag_production
 
+    def gitflow_cleanup_tags
+      return if fetch(:gitflow_keep_tags).nil?
+      tags = `git for-each-ref --format='%(*committerdate:raw)%(committerdate:raw) %(refname)' refs/tags |   sort -n  | awk '{print $3}'`
```
The `sort -n` has a little extra whitespace around it that can be removed.

bogdanRada commented:
I updated it accordingly. Let me know if it looks better now :) You were right. i forgot they accept a list of arguments, i thought initially that only one tag can be deleted at a time.

bogdanRada commented:
@technicalpickles i have been thinking...this task will delete other tags that are not generated by this gem . Maybe i should only delete the tags that are generated by this gem? Just wanted to make sure we take this in consideration also. Let me know what you think. Thank you very much.

bogdanRada commented:
@technicalpickles , any thoughts on this? Is this ok to merge ? or should i update the task to delete only tags generated by this gem? That might be difficult though if exists other tags that have same pattern but were created or from other source.

technicalpickles commented:
@bogdanRada you bring up a great point about deleting tags not created by this gem. As to whether it should be merged, I would ask you how you would feel if you had a project using this with tags, and then this gem went and deleted those unrelated tags? I know I would be pretty upset.

That tells me it should not be merged as is. I agree it'd be possible for other tags to have the same pattern that were manually created, but it's a smaller surface area than deleting all other tags.

Given there are some edge cases, this shouldn't be enabled by default (which I think it is already). There should be some documentation explaining how it works, and what considerations you'd have to make when enabling it (ie letting it own the pattern of tags).

bogdanRada commented:
Let me know if this works. i modified a bit the way the list of tags is retrieved because lightweight tags don't have date associated with it.  This is happening with "production-*" tags which seem to not be annotated.

bogdanRada commented:
i did the changes . I also modified the way i was getting the list of tags. It seems "production-*" tags are lightweight tags wihout any annotation. Not sure why though. I am using git 2.7.1 . 
And the previous way was not good because it was ignoring this tags, and were not showing in the list of tags 

I also tried to use git tag and sort by author date, but that made production and staging tags to belong to same commit and when sorting the staging tag appeared after the production tag. 

In order to preserve the correct order of tags, i eventually went with git log 
 which returns lines like this:
```
1455280667 tag: staging-2016-02-12-1-raul-10-dot-17-dot-7, tag: production-2016-02-12-1-raul-10-dot-17-dot-7, tag: invoices_iterator, tag: 10.17.7
```
 so i am sorting by the the first column ( which is timestamp} , then i am removing the first column, replace '," with new line, so that each tag will be printed on new line and replace "tag: " with empty string and remove whitespaces. Only after that i use the grep to search for tags that match the correct pattern. 


This is the only way i could find to preserve the order of tags. 

Let me know if it is ok. I also modified the README.  Thank you very much.

bogdanRada commented:
@technicalpickles , i think this is now ready for a new review. Please let me know what you think . I tested this many times and now the order of tags are preserved. I know this is a dirty solution but i couldn't find any other easier way of doing this.

technicalpickles commented on README.rdoc:
```diff
@@ -73,6 +73,14 @@ gitflow will automatically:
+=== CLEANING OLD TAGS:
+Whenever you want to clean the old tags generated by this gem you can simply do
+
+    bundle exec cap <stage_name> gitflow:cleanup_tags
```
This seems to be the only place that mentions `bundle exec`. Probably can leave it as `cap`, and do a followup PR if it should consistently use `bundle exec cap`.

technicalpickles commented on README.rdoc:
```diff
@@ -73,6 +73,14 @@ gitflow will automatically:
+
+    bundle exec cap <stage_name> gitflow:cleanup_tags
+
+This will delete all tags that have this pattern  [staging|production]{1}-[0-9]{4}-[0-9]{2}-[0-9]{2}\-([0-9]*)
```
Might be worth linking to the source where this is defined, just because it'd be easy for this to get out of date if the implementation changes.

technicalpickles commented on lib/capistrano/gitflow/helpers/helper.rb:
```diff
@@ -216,5 +216,25 @@ def gitflow_tag_production
 
+    def gitflow_cleanup_tags
+      return if fetch(:gitflow_keep_tags).nil?
+      tags = `git log --tags  --simplify-by-decoration --pretty="format:%at %D" | grep 'tag:' |sort -n | awk '{$1=""; print $0}' | tr "," "\n" | sed 's/tag:*//' | sed -e 's/^[ \t]*//' |grep -E "[staging|production]{1}-[0-9]{4}-[0-9]{2}-[0-9]{2}\-([0-9]*)"`
```
This is pretty tough to read, especially with the series of pipes. Would it be easier to process some of those things in ruby instead?

bogdanRada commented:
Let me know if it looks better now. Thanks

technicalpickles commented on lib/capistrano/gitflow/helpers/helper.rb:
```diff
@@ -218,15 +218,16 @@ def gitflow_tag_production
           system "git tag -d #{tags_to_delete.join(' ')}"
           tags_with_dots = tags_to_delete.map{ |tag| tag.prepend(':refs/tags/') }.join(' ')
-          system "git push origin #{tags_with_dots}"
+          system "git push rada #{tags_with_dots}"
```
Was this an unintentional change?

bogdanRada replied:
yeah, i was testing with a personal repository and didn't wanted to delete tags from origin while testing, because after each removal i had to re-fetch the original tags so i can test it again. Sorry. Will update it immediately.

bogdanRada commented:
Made the changes after the review. Let me know if is better now .

bogdanRada commented:
@technicalpickles , do you think any other changes are needed for this? Or is it safe to merge this pull request?

technicalpickles commented:
I can't think of anything else.

bogdanRada commented:
awesome, will merge the changes then into master and release a new minor version. Thank you very much.

bogdanRada commented:
merged the changes and release version 1.5.6 to rubygems. org. Thank you very much @technicalpickles  for your time and patience to review these changes.

zocoi commented on lib/capistrano/gitflow/helpers/helper.rb:
```diff
@@ -216,5 +216,26 @@ def gitflow_tag_production
+    def gitflow_cleanup_tags
+      return if fetch(:gitflow_keep_tags).nil?
+      tags = `git log --tags  --pretty="format:%at %D" | grep 'tag:' |sort -n | awk '{$1=""; print $0}' | tr "," "\n"| sed 's/tag:*//' | sed -e 's/^[ \t]*//'`
+      tags = tags.split.reject{|tag| tag.blank?  }
```
I think there's a bug here. When I run `cap staging gitflow:cleanup_tags`, Cap task doesn't load Rails, so I got
`NoMethodError: undefined method `blank?' for "staging-<tag>":String`

I need to add this hack
```
class String
  def blank?
    respond_to?(:empty?) ? !!empty? : !self
  end
end
```

to make it work. Can you help fixing it?

bogdanRada replied:
Fixed in this commit: https://github.com/technicalpickles/capistrano-gitflow/commit/df73e4fecb0544a9290ef89c80ae5e8551362d45 

And released version 1.5.7.
